src/lib.ml

```diff
@@ -777,16 +777,16 @@ and closure_with_overlap_checks db ts ~stack ~linking =
       Dep_stack.push stack (to_id t) >>= fun stack ->
       t.requires >>= fun deps ->
       Result.List.iter deps ~f:(loop ~stack) >>| fun () ->
-      res := t :: !res
+      res := (stack, t) :: !res
   in
   Result.List.iter ts ~f:(loop ~stack) >>= fun () ->
   let closure = List.rev !res in
-  build_impl_map closure >>= fun impls ->
+  build_impl_map closure ~stack >>= fun impls ->
   if linking && not (Lib_name.Map.is_empty impls) then
     ensure_impl_for_every_vlibs ~impls >>= fun impls ->
-    second_step_closure closure ~impls
+    second_step_closure (List.map ~f:snd closure) ~impls
```

Comment:
Should we keep the stacks here? I believe `second_step_closure` can raise.

Reply:
How can it raise? I checked again an I don't see how. `ensure_impl_for_every_vlibs` can raise and we should have some stack information there though. I'll do that.